A50-xds-outlier-detection.md

```diff
@@ -0,0 +1,262 @@
+
+The `xds_cluster_resolver` LB policy will configure an `outlier_detection` LB policy instance for each priority as the top-level LB policy for that priority.  The configuration for that `outlier_detection` policy will be based on the `outlier_detection` field in the corresponding `DiscoveryMechanism` config for that priority.
+
+In the `cds` LB policy, if the `outlier_detection` field is not set in the `Cluster` resource, a "no-op" `outlier_detection` config will be generated in the corresponding `DiscoveryMechanism` config, with `interval` set to the maximum possible value and all other fields unset.
+
+If the `outlier_detection` field *is* set in the `Cluster` resource, if the `enforcement_percentage` field is set to 0, the config `success_rate_ejection` field will be `null` and all `success_rate_*` fields will be ignored. If the `enforcing_failure_percent` field is set to 0 or `null`, the config `failure_percent_ejection` field will be `null` and all `failure_percent_*` fields will be ignored. Then the message fields will be mapped to config fields as follows:
+
+ - `interval` -> `interval`
+ - `base_ejection_time` -> `base_ejection_time`
+ - `max_ejection_time` -> `max_ejection_time`
+ - `max_ejection_percent` -> `max_ejection_percent`
+ - `success_rate_stdev_factor` -> `success_rate_ejection.stdev_factor`
+ - `enforcement_percentage` -> `success_rate_ejection.enforcement_percentage`
+ - `success_rate_minimum_hosts` -> `success_rate_ejection.minimum_hosts`
+ - `success_rate_request_volume` -> `success_rate_ejection.request_volume`
+ - `failure_percentage_threshold` -> `failure_percentage_ejection.threshold`
+ - `enforcement_percentage` -> `failure_percentage_ejection.enforcement_percentage`
+ - `failure_percentage_minimum_hosts` -> `failure_percentage_ejection.minimum_hosts`
+ - `failure_percentage_request_volume` -> `failure_percentage_ejection.request_volume`
+
+All other fields will be ignored, and all unsupported algorithms will be treated as disabled.
+
+The `child_policy` config will be the `xds_cluster_impl` policy config that previously would have been at the top level for the priority. For example, when using round robin, the LB policy tree will look like this:
+
+![gRPC Client Architecture Diagram](A50_graphics/grpc_xds_client_architecture.png)
+
+[Link to SVG file](A50_graphics/grpc_xds_client_architecture.svg)
+
+
+#### Validation
+
+The `google.protobuf.Duration` fields `interval`, `base_ejection_time`, and `max_ejection_time` must obey the restrictions in the [`google.protobuf.Duration` documentation](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#google.protobuf.Duration) and they must have non-negative values. The fields `max_ejection_percent`, `enforcement_percentage`, `failure_percentage_threshold`, and `enforcement_percentage` must have values less than or equal to `100`. If any of these requirements is violated, the `Cluster` resource should be NACKed.
```

Comment:
That is from the xDS perspective. This is marked as a general-use policy. What is the behavior when parsing a service config?

Reply:
I created a separate validation section for the non-xDS case.